src/ui/public/directives/kbn_top_nav.js

```diff
@@ -79,22 +81,34 @@ module.directive('kbnTopNav', function (Private) {
         } else {
           ctrlObj.curr = which;
         }
+
         const templateToCompile = ctrlObj.templates[ctrlObj.curr] || false;
-        $scope.kbnTopNav.currTemplate = templateToCompile ? $compile(templateToCompile)($scope) : false;
+
+        if ($scope.kbnTopNav.currTemplate) {
+          $scope.kbnTopNav.currTemplateScope.$destroy();
+          $scope.kbnTopNav.currTemplate.remove();
+        }
+
+        if (templateToCompile) {
+          $scope.kbnTopNav.currTemplateScope = $scope.$new();
```

Comment:
So then perhaps should this be further up the logic? and create a new scope for the whole directive, and clean it up on destroy. I mean wouldn't the `kbnTopNav` object stick around and cause the same problem?